acceptable/mocks.py

```diff
@@ -0,0 +1,461 @@
+            if error_list:
+                raise AssertionError(
+                    VALIDATION_ERROR_TEXT.format(
+                        data_source_name,
+                        body,
+                        schema,
+                        self._service_name,
+                        self._name,
+                        self._url,
+                        error_list,
+                    )
+                )
+
+    def _validate_request(self, request):
+        self._validate("request data", request.body, self._request_schema)
+
+    def _validate_response(self, response_body):
+        self._validate("response data", response_body, self._response_schema)
+
+    def _record_response(
+        self, request, response_status, response_headers, response_body
+    ):
+        # Shenanigans to get a response object like responses would
+        # record in calls list
+        def tmp_callback(request):
+            return response_status, response_headers, response_body
+
+        callback_response = responses.CallbackResponse(
+            request.method, self._url, tmp_callback
+        )
+        response = callback_response.get_response(request)
+        self._call_recorder.record(self, request, response, None)
+
+    def _callback(self, request):
+        try:
+            self._validate_request(request)
+            response_status, response_headers, response_body = self._response_callback(
+                request
+            )
+            self._validate_response(response_body)
+        except Exception as exc:
+            self._call_recorder.record(self, request, None, exc)
+            raise exc
+        else:
+            self._record_response(
+                request, response_status, response_headers, response_body
+            )
+            return response_status, response_headers, response_body
+
+    def get_calls(self):
+        return self._call_recorder.get_calls_for(self)
+
+    def get_last_call(self):
+        return self.get_calls()[-1]
+
+    def get_calls_matching(self, pattern):
+        return self._call_recorder.get_calls_for_matching(self, pattern)
+
+
+class EndpointMockContextManager(object):
+    def __init__(
+        self,
+        methods,
+        call_recorder,
+        service_name,
+        name,
+        url,
+        request_schema,
+        response_schema,
+        response_callback,
+    ):
+        self._methods = methods
+        self._mock = EndpointMock(
+            call_recorder,
+            service_name,
+            name,
+            methods,
+            url,
+            request_schema,
+            response_schema,
+            response_callback,
+        )
+
+    def _start(self):
+        _responses_manager.attach_callback(
+            self._methods, self._mock._url, self._mock._callback
+        )
+
+    def _stop(self):
+        _responses_manager.detach_callback(
+            self._methods, self._mock._url, self._mock._callback
+        )
```

Comment:
Again, why not just .start() and .stop()? What does the underscore gain us - it is a public api really?

Reply:
these are not public